pkg/terminal/proxy.go

```diff
@@ -0,0 +1,209 @@
+	if err != nil {
+		http.Error(w, "Failed to get the requested workspace. Cause: "+err.Error(), http.StatusForbidden)
+		return
+	}
+
+	creator := ws.GetLabels()["org.eclipse.che.workspace/creator"]
+	if creator != userId {
+		http.Error(w, "User is not a owner of the requested workspace", http.StatusForbidden)
+		return
+	}
+
+	terminalHost, err := p.getBaseTerminalHost(ws)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	p.proxyToWorkspace(terminalHost, path, user.Token, r, w)
+}
+
+func (p *Proxy) HandleProxyEnabled(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		w.Header().Set("Allow", "GET")
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	enabled, err := workspaceOperatorIsRunning()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if !enabled {
+		w.WriteHeader(http.StatusServiceUnavailable)
+		return
+	}
+	w.WriteHeader(http.StatusNoContent)
+	return
+}
+
+// stripTerminalAPIPrefix strips path prefix that is expected for Terminal API request
+func stripTerminalAPIPrefix(requestPath string) (ok bool, namespace string, workspaceName string, path string) {
+	// URL is supposed to have the following format
+	// ->   /api/terminal/proxy/{namespace}/{workspace-name}/{path} < optional
+	// -> 0 / 1 /    2   /  3  /    4      /        5       /  6
+	segments := strings.SplitN(requestPath, "/", 7)
+	if len(segments) < 6 {
+		return false, "", "", ""
+	} else {
+		namespace = segments[4]
+		workspaceName = segments[5]
+		if len(segments) == 7 {
+			path = segments[6]
+		}
+		return true, namespace, workspaceName, path
+	}
+}
+
+// getBaseTerminalHost evaluates ideUrl from the specified workspace and extract host from it
+func (p *Proxy) getBaseTerminalHost(ws *unstructured.Unstructured) (*url.URL, error) {
+	ideUrl, success, err := unstructured.NestedString(ws.UnstructuredContent(), "status", "ideUrl")
+	if !success {
+		return nil, errors.New("the specified workspace does not have ideUrl in its status")
+	}
+	if err != nil {
+		return nil, errors.New("failed to evaluate ide URL for the specified workspace. Cause: " + err.Error())
+	}
+
+	terminalUrl, err := url.Parse(ideUrl)
+	if err != nil {
+		return nil, errors.New("Failed to parse workspace ideUrl " + ideUrl)
+	}
+
+	terminalHost, err := url.Parse(terminalUrl.Scheme + "://" + terminalUrl.Host)
+	if err != nil {
+		return nil, errors.New("Failed to parse workspace ideUrl host " + ideUrl)
+	}
+
+	return terminalHost, nil
+}
+
+func (p *Proxy) proxyToWorkspace(workspaceIdeHost *url.URL, path string, token string, r *http.Request, w http.ResponseWriter) {
+	r2 := new(http.Request)
+	*r2 = *r
+	r2.URL = new(url.URL)
+	*r2.URL = *r.URL
+
+	r2.Header.Set("X-Forwarded-Access-Token", token)
+
+	r2.URL.Path = path
+
+	// TODO a new proxy per request is created. Must be revised and probably changed
+	terminalProxy := proxy.NewProxy(&proxy.Config{
```

Comment:
The requests currently proxied to the workspace are
- `.../exec/init` to initialize a terminal
- `.../activity/tick` to allow scale-to-zero after inactivity -- pings every minute: https://github.com/openshift/console/pull/5546

We can probably just make those requests directly. I think the proxy may currently be design cruft from our initial goal of serving the proxied console frontend.

Removing the proxy may interfere with future plans to reuse some of the features provided by the front-end (e.g. saving history when reopening). @sleshchenko would have to confirm on that front.

Reply:
`request.Clone` instead?

Reply:
if its relevant at this point.

Reply:
I looked into writing out such a solution and feel like it introduces some unnecessary complexity and room for error (we'd basically be writing a simplified implementation of reverseProxy.ServeHTTP at a certain point). If there are no significant objections, I'd prefer to keep it as-is for now and potentially look into it in the future.

Regarding the security concerns:
- It's likely we'll have additional API paths on the workspace side for 4.6, as we intend to look into reusing some of the cloudshell UI
- Maliciously proxying the user's token to different paths is not a significant concern since anything malicious you could do with an alternate path you could do with the default path (e.g. log token on `/activity/tick`). If we want to restrict API paths, it'd be easier to check against a list of "approved" paths.

There may be some headers/other data that is processed by reverseProxy that's problematic that I'm unaware of. If it's important I can work on this today and have changes by tomorrow. @benjaminapetersen @spadgett WDYT?

Reply:
Do we need the token for these other requests? I'd expect only to need the token once on init.

We should only pass the token when we need to.

> There may be some headers/other data that is processed by reverseProxy that's problematic that I'm unaware of.

It's very likely. We should not be passing cookies for instance. Probably some other headers could cause problems.

If there are only a handful of requests we'll need to make, it would be better to support only those explicitly. The proxy seems unnecessary to me here. I'd think removing the proxy would *reduce* complexity as opposed to adding to it.

Reply:
@spadgett makes sense to me, I'll work on it.

Reply:
Implemented, PTAL.